File: packages/camera/camera_android_camerax/test/camera_selector_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:camera_android_camerax/src/camera_info.dart';
import 'package:camera_android_camerax/src/camera_selector.dart';
import 'package:camera_android_camerax/src/camerax_library.g.dart';
import 'package:camera_android_camerax/src/instance_manager.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';

import 'camera_selector_test.mocks.dart';
import 'test_camerax_library.g.dart';

@GenerateMocks(<Type>[TestCameraSelectorHostApi])
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('CameraSelector', () {
    tearDown(() => TestCameraSelectorHostApi.setup(null));

    test('detachedCreateTest', () async {
      final MockTestCameraSelectorHostApi mockApi =
          MockTestCameraSelectorHostApi();
      TestCameraSelectorHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      CameraSelector.detached(
        instanceManager: instanceManager,
      );

      verifyNever(mockApi.create(argThat(isA<int>()), null));
    });

    test('createTestWithoutLensSpecified', () async {
      final MockTestCameraSelectorHostApi mockApi =
          MockTestCameraSelectorHostApi();
      TestCameraSelectorHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      CameraSelector(
        instanceManager: instanceManager,
      );

      verify(mockApi.create(argThat(isA<int>()), null));
    });

    test('createTestWithLensSpecified', () async {
      final MockTestCameraSelectorHostApi mockApi =
          MockTestCameraSelectorHostApi();
      TestCameraSelectorHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      CameraSelector(
          instanceManager: instanceManager,
          lensFacing: CameraSelector.lensFacingBack);

      verify(
          mockApi.create(argThat(isA<int>()), CameraSelector.lensFacingBack));
    });

    test('filterTest', () async {
      final MockTestCameraSelectorHostApi mockApi =
          MockTestCameraSelectorHostApi();
      TestCameraSelectorHostApi.setup(mockApi);

      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      final CameraSelector cameraSelector = CameraSelector.detached(
        instanceManager: instanceManager,
      );
      const int cameraInfoId = 3;
      final CameraInfo cameraInfo =
          CameraInfo.detached(instanceManager: instanceManager);

      instanceManager.addHostCreatedInstance(
        cameraSelector,
        0,
        onCopy: (_) => CameraSelector.detached(),
      );
      instanceManager.addHostCreatedInstance(
        cameraInfo,
        cameraInfoId,
        onCopy: (_) => CameraInfo.detached(),
      );

      when(mockApi.filter(instanceManager.getIdentifier(cameraSelector),
          <int>[cameraInfoId])).thenReturn(<int>[cameraInfoId]);
      expect(await cameraSelector.filter(<CameraInfo>[cameraInfo]),
          equals(<CameraInfo>[cameraInfo]));

      verify(mockApi.filter(0, <int>[cameraInfoId]));
    });

    test('flutterApiCreateTest', () {
      final InstanceManager instanceManager = InstanceManager(
        onWeakReferenceRemoved: (_) {},
      );
      final CameraSelectorFlutterApi flutterApi = CameraSelectorFlutterApiImpl(
        instanceManager: instanceManager,
      );

      flutterApi.create(0, CameraSelector.lensFacingBack);

      expect(instanceManager.getInstanceWithWeakReference(0),
          isA<CameraSelector>());
      expect(
          (instanceManager.getInstanceWithWeakReference(0)! as CameraSelector)
              .lensFacing,
          equals(CameraSelector.lensFacingBack));
    });
  });
}

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraSelectorTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camerax;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraSelector;
import io.flutter.plugin.common.BinaryMessenger;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

public class CameraSelectorTest {
  @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();

  @Mock public CameraSelector mockCameraSelector;
  @Mock public BinaryMessenger mockBinaryMessenger;

  InstanceManager testInstanceManager;

  @Before
  public void setUp() {
    testInstanceManager = InstanceManager.open(identifier -> {});
  }

  @After
  public void tearDown() {
    testInstanceManager.close();
  }

  @Test
  public void createTest() {
    final CameraSelectorHostApiImpl cameraSelectorHostApi =
        new CameraSelectorHostApiImpl(mockBinaryMessenger, testInstanceManager);
    final CameraXProxy mockCameraXProxy = mock(CameraXProxy.class);
    final CameraSelector.Builder mockCameraSelectorBuilder = mock(CameraSelector.Builder.class);

    cameraSelectorHostApi.cameraXProxy = mockCameraXProxy;
    when(mockCameraXProxy.createCameraSelectorBuilder()).thenReturn(mockCameraSelectorBuilder);

    when(mockCameraSelectorBuilder.requireLensFacing(1)).thenReturn(mockCameraSelectorBuilder);
    when(mockCameraSelectorBuilder.build()).thenReturn(mockCameraSelector);

    cameraSelectorHostApi.create(0L, 1L);

    verify(mockCameraSelectorBuilder).requireLensFacing(CameraSelector.LENS_FACING_BACK);
    assertEquals(testInstanceManager.getInstance(0L), mockCameraSelector);
  }

  @Test
  public void filterTest() {
    final CameraSelectorHostApiImpl cameraSelectorHostApi =
        new CameraSelectorHostApiImpl(mockBinaryMessenger, testInstanceManager);
    final CameraInfo cameraInfo = mock(CameraInfo.class);
    final List<CameraInfo> cameraInfosForFilter = Arrays.asList(cameraInfo);
    final List<Long> cameraInfosIds = Arrays.asList(1L);

    testInstanceManager.addDartCreatedInstance(mockCameraSelector, 0);
    testInstanceManager.addDartCreatedInstance(cameraInfo, 1);

    when(mockCameraSelector.filter(cameraInfosForFilter)).thenReturn(cameraInfosForFilter);

    assertEquals(
        cameraSelectorHostApi.filter(0L, cameraInfosIds),
        Arrays.asList(testInstanceManager.getIdentifierForStrongReference(cameraInfo)));
    verify(mockCameraSelector).filter(cameraInfosForFilter);
  }

  @Test
  public void flutterApiCreateTest() {
    final CameraSelectorFlutterApiImpl spyFlutterApi =
        spy(new CameraSelectorFlutterApiImpl(mockBinaryMessenger, testInstanceManager));

    spyFlutterApi.create(mockCameraSelector, 0L, reply -> {});

    final long identifier =
        Objects.requireNonNull(
            testInstanceManager.getIdentifierForStrongReference(mockCameraSelector));
    verify(spyFlutterApi).create(eq(identifier), eq(0L), any());
  }
}
